quickwit-search/src/fetch_docs.rs

```diff
@@ -144,30 +163,140 @@ async fn get_searcher_for_split_without_cache(
             let doc = searcher
                 .doc_async(global_doc_addr.doc_addr)
                 .await
                 .context("searcher-doc-async")?;
             let doc_json = searcher.schema().to_json(&doc);
-            Ok((global_doc_addr, doc_json))
+            if moved_snippet_generators.is_empty() {
+                return Ok((
+                    global_doc_addr,
+                    DocBundle {
+                        doc_json,
+                        highlights_json: None,
+                    },
+                ));
+            }
+
+            let mut highlight = HashMap::new();
+            for (field, field_values) in doc.get_sorted_field_values() {
+                let field_name = searcher.schema().get_field_name(field);
+                if let Some(snippet_generator) = moved_snippet_generators.get(field_name) {
+                    let values = snippets_from_field_values(snippet_generator, field_values);
+                    highlight.insert(field_name, values);
+                }
+            }
+            let highlight_json = serde_json::to_string(&highlight)?;
+            Ok((
+                global_doc_addr,
+                DocBundle {
+                    doc_json,
+                    highlights_json: Some(highlight_json),
+                },
+            ))
         }
     });
 
     let stream = futures::stream::iter(doc_futures).buffer_unordered(NUM_CONCURRENT_REQUESTS);
     stream.try_collect::<Vec<_>>().await
 }
+
+// Creates the snippets generators associated to the snippet fields
+// from a search request.
+async fn create_snippet_generators(
+    searcher: &Searcher,
+    doc_mapper: Arc<dyn DocMapper>,
+    search_request: &SearchRequest,
+) -> anyhow::Result<HashMap<String, SnippetGenerator>> {
+    let schema = searcher.schema();
+    let query = doc_mapper.query(schema.clone(), search_request)?;
+
+    let mut snippet_generators = HashMap::new();
+    for field_name in &search_request.snippet_fields {
+        let field = schema
+            .get_field(field_name)
+            .ok_or_else(|| QueryParserError::FieldDoesNotExist(field_name.clone()))?;
+        let snippet_generator = create_snippet_generator(searcher, &*query, field).await?;
+        snippet_generators.insert(field_name.clone(), snippet_generator);
+    }
+    Ok(snippet_generators)
+}
+
+// Creates a snippet generator associated to a field.
+async fn create_snippet_generator(
+    searcher: &Searcher,
+    query: &dyn Query,
+    field: Field,
+) -> anyhow::Result<SnippetGenerator> {
+    let mut terms = BTreeMap::new();
+    query.query_terms(&mut terms);
+    let mut terms_text: BTreeMap<String, Score> = Default::default();
+    for (term, _) in terms {
+        if term.field() != field {
+            continue;
+        }
+        let term_str = if let Some(term_str) = term.as_str() {
+            term_str
+        } else {
+            continue;
+        };
+        let doc_freq = searcher.doc_freq_async(&term).await?;
```

Comment:
Should we return the doc frequencies in partial hits to avoid another fetch of the term dictionary during the fetch docs phase?

Reply:
Not easy to do :/
We have a frequency per term and this would mean adding a list of (term, freq) for each partial hit. On possibility would be to retrieve from each leaf request the terms freq and put it then in the fetch_docs query but it seems too complex.

Reply:
Here what we can do is warming the freq before calling tantivy snippet. I believe this will avoid the copy paste from tantivy code.

Reply:
yeah... let's just eventually get better at caching.